python/GafferUI/Menu.py

```diff
@@ -462,24 +463,47 @@ def __updateSearchMenu( self, text ) :
 			self.__searchLine.setFocus()
 			self.__searchMenu.setUpdatesEnabled( True )
 
-	def __matchingActions( self, searchText ) :
+	def __matchingActions( self, searchText, noMoreMatches=False ) :
 
+		searchText = searchText[:99]  # re matching only supports up to 100 groups
 		results = {}
-		matcher = re.compile( re.escape( searchText ), re.IGNORECASE )
+
+		# search and group all in between characters. e.g: (.*)t(.*)e(.*)s(.*)t
+		searchStrings = [ re.escape(searchText[i]) for i in range( len( searchText ) ) ]
+		regex = '(.*)' + '(.*)'.join( searchStrings )
+		matcher = re.compile( regex, re.IGNORECASE )
 
 		for name in self.__searchStructure :
 
 			match = matcher.search( name )
+
 			if match :
 
+				# weight by how many sub-strings we get when splitting at the match positions
+				weight = len( [ e for i,e in enumerate( match.groups() ) if e and i > 0 ] )
+
+				# position of the first matching character
+				pos = len( match.group(1) )
+
 				for item, path in self.__searchStructure[name] :
 
 					action = self.__buildAction( item, name, self.__searchMenu )
 					if name not in results :
-						results[name] = { "pos" : match.start(), "actions" : [] }
+						results[name] = { "pos" : pos, "weight" : weight, "actions" : [] }
 
 					results[name]["actions"].append( ( action, path ) )
 
+		if not results and len( searchText ) > 1:
+
+			self.__searchLine.setProperty( "gafferError", GafferUI._Variant.toVariant( True ) )
```

Comment:
This seems like a poor separation of concerns - I wouldn't expect `__matchingActions()` to have side effects. The previous code split the concerns fairly neatly so that `__matchingActions()` performed the search, and `__updateSearchMenu()` was responsible for displaying the results. Can we keep this structure?

Reply:
for sure!